File: gymnasium/lib/gymnasion_grpc/dimensions/project.ex
defmodule GymnasiumGrpc.Dimensions.Project do
  alias GymnasiumGrpc.Helpers
  alias Gymnasium.{Dimensions.Project, Dimensions}
  alias Proto.Gymnasium.V1.Dimensions.Project, as: ProtoProject

  alias Proto.Gymnasium.V1.{
    ListProjectRecordsResponse,
    CreateProjectRecordResponse,
    GetProjectRecordResponse,
    ArchiveProjectRecordResponse,
    RestoreProjectRecordResponse,
    DeleteProjectRecordResponse
  }

  def list(attrs \\ []) do
    project_records =
      attrs
      |> Enum.into(%{
        archived: false
      })
      |> Dimensions.list_projects()
      |> Enum.map(fn project -> to_proto_project(project) end)

    %ListProjectRecordsResponse{
      project_records: project_records
    }
  end

  def create(attrs \\ %{}) do
    {:ok, project} = Dimensions.create_project(attrs)

    %CreateProjectRecordResponse{
      project_record: to_proto_project(project)
    }
  end

  def find(slug) do
    project_record =
      slug
      |> Dimensions.find_project!()
      |> to_proto_project

    %GetProjectRecordResponse{
      project_record: project_record
    }
  end

  def archive(id) do
    project = Dimensions.get_project!(id)

    {:ok, project} = Dimensions.archive_project(project)

    %ArchiveProjectRecordResponse{
      project_record: to_proto_project(project)
    }
  end

  def restore(id) do
    project = Dimensions.get_project!(id)

    {:ok, project} = Dimensions.restore_project(project)

    %RestoreProjectRecordResponse{
      project_record: to_proto_project(project)
    }
  end

  def delete(id) do
    project = Dimensions.get_project!(id)

    {:ok, _project} = Dimensions.delete_project(project)

    %DeleteProjectRecordResponse{}
  end

  defp to_proto_project(project = %Project{}) do
    archive_time = if project.archived_at, do: Helpers.to_proto_timestamp(project.archived_at)

    %ProtoProject{
      id: project.id,
      archive_time: archive_time,
      description: project.description,
      name: project.name,
      slug: project.slug,
      create_time: Helpers.to_proto_timestamp(project.inserted_at),
      update_time: Helpers.to_proto_timestamp(project.updated_at)
    }
  end
end

File: temple2/src/datastore/project.rs
use crate::{
    datastore::{self, Repo},
    model::{Project, Uuid},
    util, AppResult,
};

pub struct ListProjectsAttributes {
    pub archived: bool,
}

pub async fn list(repo: &Repo, attributes: ListProjectsAttributes) -> AppResult<Vec<Project>> {
    let ListProjectsAttributes { archived } = attributes;

    let projects = repo
        .select_projects(datastore::repo::SelectProjectsAttributes { archived })
        .await?;

    Ok(projects)
}

pub struct CreateProjectAttributes {
    pub name: String,
    pub description: String,
}

pub async fn create(repo: &Repo, attributes: CreateProjectAttributes) -> AppResult<Project> {
    let CreateProjectAttributes { name, description } = attributes;

    let project = repo
        .create_project(Project {
            archivation_time: Default::default(),
            create_time: Default::default(),
            description,
            id: Default::default(),
            slug: util::slug::sluggify(&name),
            name,
        })
        .await?;

    Ok(project)
}

pub async fn get(repo: &Repo, id: Uuid) -> AppResult<Project> {
    let project = repo.get_project(id).await?;

    Ok(project)
}

pub async fn find(repo: &Repo, slug: &str) -> AppResult<Project> {
    let project = repo.find_project(slug).await?;

    Ok(project)
}

pub async fn update(repo: &Repo, project: Project) -> AppResult<Project> {
    let project = repo.update_project(project).await?;

    Ok(project)
}

pub async fn delete(repo: &Repo, project: Project) -> AppResult<()> {
    repo.delete_project(project).await
}
